### PR TITLE
ROX-17328: networflows offline v1

### PR DESCRIPTION
**Pull request by lvalerom**
## Description

OfflineV1 implementation of the NetworkFlows component. The NetworkFlows pipeline is divide in two; the receiving/storing of NetworkFlows from collector, and the processing/sending of these NetworkFlows. When we switch to offline mode we only stop the second part of the pipeline, meaning we still receive flows from collector and we save them in our internal caches.

## Checklist
- [x] Investigated and inspected CI test results
- [x] Unit test and regression tests added
- ~[ ] Evaluated and added CHANGELOG entry if required~
- ~[ ] Determined and documented upgrade steps~
- ~[ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))~

If any of these don't apply, please comment below.

## Testing Performed

- [x] Unit tests
- [x] Manual testing
    1. Deploy ACS in a cluster with `ROX_PREVENT_SENSOR_RESTART_ON_DISCONNECT` set to `true`
    2. Create a nginx deployment and service to connect to it
    3. Scale down central to break the connection with sensor
    4. Check that sensor is in the retry loop
    5. Create a pod (e.g. `talk`) that connects to the nginx deployment (e.g. `kubectl -n test run talk --rm -i --tty --image quay.io/rhacs-demo/netflow -- -connect nginx-service.test.svc.cluster.local:80`)
    6. Scale up central and stop the pod (`talk`) that connects to the nginx
    7. Check that we can see a flow to the nginx deployment in the NetworkGraph in central
- [x] CI

### CODE REVIEW

File: sensor/common/networkflow/manager/manager_impl_test.go [hunk 1]
```diff
@@ -179,7 +193,521 @@ func (suite *NetworkflowManagerTestSuite) TestAddNoOriginator() {
...
+		expectEntityLookupContainer expectFn
+		expectEntityLookupEndpoint  expectFn
+		expectExternalLookup        expectFn
+		expectedIndicator           *networkConnIndicator
+		expectedConnection          *connection
+		expectedStatus              *connStatus
+	}{
+		"Rotten connection": {
+			connPair:                    createConnectionPair(true, true, false, timestamp.Now().Add(-maxContainerResolutionWaitPeriod*2)),
+			expectEntityLookupContainer: expectEntityLookupContainerHelper(mockEntityStore, 1, clusterentities.ContainerMetadata{}, false),
```
Comment (vikin91) [inline sensor/common/networkflow/manager/manager_impl_test.go:221]: I am wondering whether the code without helpers would look better?

```suggestion
			expectEntityLookupContainer: func() {
				mockEntityStore.EXPECT().LookupByContainerID(gomock.Any()).Times(1).DoAndReturn(func(_ any) (clusterentities.ContainerMetadata, bool) {
					return clusterentities.ContainerMetadata{}, false
				})
			}
```

File: sensor/common/networkflow/manager/manager_impl_test.go [hunk 2]
```diff
@@ -179,7 +193,521 @@ func (suite *NetworkflowManagerTestSuite) TestAddNoOriginator() {
...
+
+//region networkFlowManager tests
+
+func (s *NetworkFlowManagerTestSuite) TestEnrichConnection() {
+	mockCtrl := gomock.NewController(s.T())
+	m, mockEntityStore, mockExternalSrc, _ := createManager(mockCtrl)
+	containerID := "id"
+	cases := map[string]struct {
+		connPair                    *connectionPair
+		enrichConnections           map[networkConnIndicator]timestamp.MicroTS
```
Comment (vikin91) [inline sensor/common/networkflow/manager/manager_impl_test.go:211]:
```suggestion
		enrichedConnections           map[networkConnIndicator]timestamp.MicroTS
```

File: sensor/common/networkflow/manager/manager_impl_test.go [hunk 3]
```diff
@@ -179,7 +193,521 @@ func (suite *NetworkflowManagerTestSuite) TestAddNoOriginator() {
...
+		mockDetector.EXPECT().ProcessNetworkFlow(gomock.Any()).Times(times)
+	}
+}
+
+type connectionPair struct {
+	conn   *connection
+	status *connStatus
+}
+
+func createConnectionPair(incoming bool, external bool, invalid bool, firstSeen timestamp.MicroTS) *connectionPair {
```
Comment (vikin91) [inline sensor/common/networkflow/manager/manager_impl_test.go:601]: This may be an overkill here, but applying a builder pattern would make it easier to parse the 3 bools when reading the test cases.

File: sensor/common/networkflow/manager/manager_impl_test.go [hunk 4]
```diff
@@ -179,7 +193,521 @@ func (suite *NetworkflowManagerTestSuite) TestAddNoOriginator() {
...
+		expectedStatus              *connStatus
+	}{
+		"Rotten connection": {
+			connPair:                    createConnectionPair(true, true, false, timestamp.Now().Add(-maxContainerResolutionWaitPeriod*2)),
+			expectEntityLookupContainer: expectEntityLookupContainerHelper(mockEntityStore, 1, clusterentities.ContainerMetadata{}, false),
+			expectedStatus: &connStatus{
+				rotten: true,
+			},
+		},
+		"Incoming External connection no external sources hit": {
```
Comment (vikin91) [inline sensor/common/networkflow/manager/manager_impl_test.go:226]: I am guessing that the `no` negates the `hit` - if yes, then maybe let's call it a `miss`.
Next, I think that the `miss` here is the `nil` in `expectExternalLookupHelper`. As hit and miss are mostly use for reading the cache, then maybe we could rename it to something like:

```suggestion
		"Incoming External connection should not result in external lookup": {
```

File: sensor/common/networkflow/manager/manager_impl_test.go [hunk 5]
```diff
@@ -179,7 +193,521 @@ func (suite *NetworkflowManagerTestSuite) TestAddNoOriginator() {
...
+			expectedIndicator: &networkConnIndicator{
+				dstPort:  80,
+				protocol: net.TCP.ToProtobuf(),
+				dstEntity: networkgraph.EntityFromProto(&storage.NetworkEntityInfo{
+					Id: containerID,
+				}),
+				srcEntity: networkgraph.EntityForDeployment(containerID),
+			},
+		},
+		"Incoming": {
```
Comment (vikin91) [inline sensor/common/networkflow/manager/manager_impl_test.go:264]: I think, it may be beneficial to rewrite the test case names, so that it includes the word `should`. It is not an issue when the names are long, but at least we would know what a given case is testing.

**Review by vikin91 — COMMENTED** [review]
Posting my partial review so far. Didn't make it to go through all the tests, but I will continue the next day.

File: sensor/common/networkflow/manager/manager_impl.go [hunk 6]
```diff
@@ -632,6 +671,8 @@ func computeUpdatedProcesses(current map[processListeningIndicator]timestamp.Mic
 		}
 	}
 
+	previousMutex.RLock()
+	defer previousMutex.RUnlock()
```
[the remark this answers is not in the document]
Comment (lvalerom) [reply sensor/common/networkflow/manager/manager_impl.go]: Oops, fixed!

File: sensor/common/networkflow/manager/manager_impl.go [hunk 7]
```diff
@@ -246,20 +252,60 @@ func (m *networkFlowManager) Capabilities() []centralsensor.SensorCapability {
...
+
+func (m *networkFlowManager) resetLastSentState() {
+	m.lastSentStateMutex.Lock()
+	defer m.lastSentStateMutex.Unlock()
+	m.enrichedConnsLastSentState = nil
+	m.enrichedEndpointsLastSentState = nil
+	m.enrichedProcessesLastSentState = nil
+}
+
+func (m *networkFlowManager) swapConnectionStates(newConns map[networkConnIndicator]timestamp.MicroTS, newEndpoints map[containerEndpointIndicator]timestamp.MicroTS) {
```
[the remark this answers is not in the document]
Comment (lvalerom) [reply sensor/common/networkflow/manager/manager_impl.go]: Makes sense. I've changed to `updateConnections`

File: sensor/common/networkflow/manager/manager_impl.go [hunk 8]
```diff
@@ -246,20 +252,60 @@ func (m *networkFlowManager) Capabilities() []centralsensor.SensorCapability {
...
+	m.lastSentStateMutex.Lock()
+	defer m.lastSentStateMutex.Unlock()
+	m.enrichedConnsLastSentState = newConns
+	m.enrichedEndpointsLastSentState = newEndpoints
+}
+
+func (m *networkFlowManager) swapProcessesState(newProcesses map[processListeningIndicator]timestamp.MicroTS) {
```
[the remark this answers is not in the document]
Comment (lvalerom) [reply sensor/common/networkflow/manager/manager_impl.go]: Same as above

File: sensor/common/networkflow/manager/manager_impl_test.go (hunk 2 above)
Comment (lvalerom) [reply sensor/common/networkflow/manager/manager_impl_test.go]: Renamed.

File: sensor/common/networkflow/manager/manager_impl_test.go (hunk 3 above)
Comment (lvalerom) [reply sensor/common/networkflow/manager/manager_impl_test.go]: I went for it and I think it's way more readable :) Let me know what you think

File: sensor/common/networkflow/manager/manager_impl_test.go (hunk 5 above)
Comment (lvalerom) [reply sensor/common/networkflow/manager/manager_impl_test.go]: I've renamed pretty much all the cases in `TestEnrichConnection`, `TestEnrichContainerEndpoint`, and `TestEnrichProcessListening`. Let me know if you think they are more understandable now.

File: sensor/common/networkflow/manager/manager_impl_test.go (hunk 4 above)
Comment (lvalerom) [reply sensor/common/networkflow/manager/manager_impl_test.go]: See comment below

File: sensor/common/networkflow/manager/manager_impl_test.go (hunk 1 above)
Comment (lvalerom) [reply sensor/common/networkflow/manager/manager_impl_test.go:222]: I did that at the beginning at it seemed a little bit difficult to read IMO. I can try to see if I like it now though :)

**Comment by vikin91** [comment]
I am looking at this now. Hope to submit a review (partial or full) within the next hour

File: sensor/common/networkflow/manager/manager_impl.go [hunk 9]
```diff
@@ -585,7 +620,7 @@ func (m *networkFlowManager) currentEnrichedProcesses() map[processListeningIndi
 	return enrichedProcesses
 }
 
-func computeUpdatedConns(current map[networkConnIndicator]timestamp.MicroTS, previous map[networkConnIndicator]timestamp.MicroTS) []*storage.NetworkFlow {
+func computeUpdatedConns(current map[networkConnIndicator]timestamp.MicroTS, previous map[networkConnIndicator]timestamp.MicroTS, previousMutest *sync.RWMutex) []*storage.NetworkFlow {
```
Comment (vikin91) [inline sensor/common/networkflow/manager/manager_impl.go:623]:
```suggestion
func computeUpdatedConns(current map[networkConnIndicator]timestamp.MicroTS, previous map[networkConnIndicator]timestamp.MicroTS, previousMutex *sync.RWMutex) []*storage.NetworkFlow {
```

File: sensor/common/networkflow/manager/manager_impl.go [hunk 10]
```diff
@@ -595,6 +630,8 @@ func computeUpdatedConns(current map[networkConnIndicator]timestamp.MicroTS, pre
 		}
 	}
 
+	previousMutest.RLock()
```
Comment (vikin91) [inline sensor/common/networkflow/manager/manager_impl.go:633]: We do `prevTS, ok := previous[conn]` in line 627 - we need to lock this one earlier.

File: sensor/common/networkflow/manager/manager_impl_test.go [hunk 11]
```diff
@@ -179,7 +194,696 @@ func (suite *NetworkflowManagerTestSuite) TestAddNoOriginator() {
...
+
+func (s *NetworkFlowManagerTestSuite) TestManagerOfflineMode() {
+	s.T().Setenv(env.ProcessesListeningOnPort.EnvVar(), "false")
+	srcID := "src-id"
+	dstID := "dst-id"
+	mockCtrl := gomock.NewController(s.T())
+	hostname := "hostname"
+	containerID := "container-id"
+	m, mockEntity, _, mockDetector := createManager(mockCtrl)
+	states := []struct {
```
Comment (vikin91) [inline sensor/common/networkflow/manager/manager_impl_test.go:461]: It would be cool to add names to the test cases, so that one can easier understand what is being tested

```suggestion
	states := map[string]struct {
```

File: sensor/common/networkflow/manager/manager_impl_test.go [hunk 12]
```diff
@@ -179,7 +194,696 @@ func (suite *NetworkflowManagerTestSuite) TestAddNoOriginator() {
...
+			}
+		})
+	}
+	m.Stop(nil)
+	m.done.Wait()
+}
+
+// endregion
+
+// region Helper functions
```
Comment (vikin91) [inline sensor/common/networkflow/manager/manager_impl_test.go:603]: This region could be moved to another file - helpers can go to the famous `testing_test.go` file.

File: sensor/common/networkflow/manager/manager_impl_test.go [hunk 13]
```diff
@@ -179,7 +194,696 @@ func (suite *NetworkflowManagerTestSuite) TestAddNoOriginator() {
...
+	return &central.MsgFromSensor{
+		Msg: &central.MsgFromSensor_NetworkFlowUpdate{
+			NetworkFlowUpdate: &central.NetworkFlowUpdate{
+				UpdatedEndpoints: updates,
+			},
+		},
+	}
+}
+
+func assertSensorMessageConnectionIDs(s *NetworkFlowManagerTestSuite, expectedUpdates []*storage.NetworkFlow, actualUpdates []*storage.NetworkFlow) {
```
Comment (vikin91) [inline sensor/common/networkflow/manager/manager_impl_test.go:862]: Why not
```suggestion
func (s *NetworkFlowManagerTestSuite) assertSensorMessageConnectionIDs(expectedUpdates []*storage.NetworkFlow, actualUpdates []*storage.NetworkFlow) {
```

?

**Review by vikin91 — COMMENTED** [review]
I finally made it through the code - It looks really good!

Let's fix the mutex locking and this will be ✅ 
You may consider also my other comments, but they are optional.

File: sensor/common/networkflow/manager/manager_impl.go (hunk 9 above)
Comment (lvalerom) [reply sensor/common/networkflow/manager/manager_impl.go]: Fixed 😅

File: sensor/common/networkflow/manager/manager_impl.go (hunk 10 above)
Comment (lvalerom) [reply sensor/common/networkflow/manager/manager_impl.go]: Good catch! I fixed the other ones and completely missed this one.

File: sensor/common/networkflow/manager/manager_impl_test.go (hunk 11 above)
Comment (lvalerom) [reply sensor/common/networkflow/manager/manager_impl_test.go:456]: This actually needs to be a slice to maintain order. We rely on the connections being added in order to later check for them in the next state.

File: sensor/common/networkflow/manager/manager_impl_test.go (hunk 12 above)
Comment (lvalerom) [reply sensor/common/networkflow/manager/manager_impl_test.go]: Moved!

File: sensor/common/networkflow/manager/manager_impl_test.go (hunk 11 above)
Comment (vikin91) [reply sensor/common/networkflow/manager/manager_impl_test.go:456]: Good point! Maybe we should write this in the comment that the order is important?
We could still add a `name string` to the struct 😉

File: sensor/common/networkflow/manager/manager_impl_test.go (hunk 3 above)
Comment (vikin91) [reply sensor/common/networkflow/manager/manager_impl_test.go]: It looks good, thanks!

File: sensor/common/networkflow/manager/manager_impl_test.go (hunk 5 above)
Comment (vikin91) [reply sensor/common/networkflow/manager/manager_impl_test.go]: Thanks for that! I feel the difference is huge :)

**Review by vikin91 — APPROVED** [review]
Thanks for improving! The code looks good to me

File: sensor/common/networkflow/manager/manager_impl_test.go (hunk 11 above)
Comment (lvalerom) [reply sensor/common/networkflow/manager/manager_impl_test.go:456]: That I can do 🙁

File: sensor/common/networkflow/manager/manager_impl_test.go (hunk 1 above)
Comment (lvalerom) [reply sensor/common/networkflow/manager/manager_impl_test.go]: I tried it in the `TestEnrichConenction` (see [ecccc60](https://github.com/stackrox/stackrox/pull/6959/commits/ecccc60f9d5c978930334fc29a8c20fcec00989a)). I don't have a strong opinion regarding this. On one hand I think it looks better because you don't have to check what the helper functions do, but on the other hand I think it is too verbose somehow. If you think it looks better I'll change it in the rest of the tests, otherwise I'll revert the last commit 🙂

File: sensor/common/networkflow/manager/manager_impl_test.go (hunk 11 above)
Comment (lvalerom) [reply sensor/common/networkflow/manager/manager_impl_test.go:513]: Done

Comment (vikin91) [reply sensor/common/networkflow/manager/manager_impl_test.go:513]: That looks great, thanks!

File: sensor/common/networkflow/manager/manager_impl_test.go (hunk 1 above)
Comment (vikin91) [reply sensor/common/networkflow/manager/manager_impl_test.go]: I don't have strong opinion, but I think it looks better without the helpers = less scrolling to understand what is going on. 
We can also ask other colleagues for opinion if you need stronger signals :)

File: sensor/common/networkflow/manager/manager_impl_test.go [hunk 14]
```diff
@@ -179,7 +189,472 @@ func (suite *NetworkflowManagerTestSuite) TestAddNoOriginator() {
...
+	dstID := "dst-id"
+	cases := map[string]struct {
+		connPair                    *connectionPair
+		enrichedConnections         map[networkConnIndicator]timestamp.MicroTS
+		expectEntityLookupContainer expectFn
+		expectEntityLookupEndpoint  expectFn
+		expectExternalLookup        expectFn
+		expectedIndicator           *networkConnIndicator
+		expectedConnection          *connection
+		expectedStatus              *connStatus
```
Comment (fredrb) [inline sensor/common/networkflow/manager/manager_impl_test.go:214]: Another bit of a nitpick here, but could we differentiate the prefix from the gomock `EXPECT()` type calls and actual assertions?

Maybe the latter three could have `assert` prefix rather than `expect`? I think the testCase struct would read much better:
```go
connPair                  *connectionPair
enrichedConnections       map[networkConnIndicator]timestamp.MicroTS
expectedCalls             expectFn
assertIndicator           *networkConnIndicator
assertConnection          *connection
assertStatus              *connStatus          
```

File: sensor/common/networkflow/manager/manager_impl_test.go [hunk 15]
```diff
@@ -179,7 +189,472 @@ func (suite *NetworkflowManagerTestSuite) TestAddNoOriginator() {
...
+	}{
+		"Rotten connection should return rotten status": {
+			connPair: createConnectionPair().incoming().external().firstSeen(timestamp.Now().Add(-maxContainerResolutionWaitPeriod * 2)),
+			expectEntityLookupContainer: func() {
+				mockEntityStore.EXPECT().LookupByContainerID(gomock.Any()).Times(1).DoAndReturn(func(_ any) (clusterentities.ContainerMetadata, bool) {
```
Comment (fredrb) [inline sensor/common/networkflow/manager/manager_impl_test.go:219]: This expectation has the same parameters and times called in every test case. Let's not define this for each test case and simplify it a bit. 

Why not have a `clusterentities.ContainerMetadata` property in the test case, called `givenContainerMetadata`. This way each test case would define its own, and for the empty case you just don't define it (as the empty value for a struct is the struct itself with all empty values).

File: sensor/common/networkflow/manager/manager_impl_test.go [hunk 16]
```diff
@@ -179,7 +189,472 @@ func (suite *NetworkflowManagerTestSuite) TestAddNoOriginator() {
...
+			connPair:            createConnectionPair().incoming().external(),
+			enrichedConnections: make(map[networkConnIndicator]timestamp.MicroTS),
+			expectEntityLookupContainer: func() {
+				mockEntityStore.EXPECT().LookupByContainerID(gomock.Any()).Times(1).DoAndReturn(func(_ any) (clusterentities.ContainerMetadata, bool) {
+					return clusterentities.ContainerMetadata{
+						DeploymentID: dstID,
+					}, true
+				})
+			},
+			expectExternalLookup: func() {
```
Comment (fredrb) [inline sensor/common/networkflow/manager/manager_impl_test.go:237]: WDYT about creating some setup functions for these calls? Something like
```go

func givenNetworkLookup(info *storage.NetworkEntityInfo) {
       mockExternalSrc.EXPECT().LookupByNetwork(gomock.Any()).Times(1).DoAndReturn(func(_ any) *storage.NetworkEntityInfo {
              return info
       })
}

// in the test case definition
expectExternalLookup: func() { givenNetworkLookup( /* define here */) }
// in another
expectExternalLookup: func() { givenNetworkLookup(nil) }
```

Just to remove a bit of the boilerplate here which makes reading the test a bit harder. You could return a function also to pass it directly
```go
expectExternalLookup: givenNetworkLookup(nil)
```

But that might conflict with the suggestion above of making a single `expect` block.

File: sensor/common/networkflow/manager/manager_impl_test.go [hunk 17]
```diff
@@ -179,7 +189,472 @@ func (suite *NetworkflowManagerTestSuite) TestAddNoOriginator() {
...
+	mockCtrl := gomock.NewController(s.T())
+	m, mockEntityStore, _, _ := createManager(mockCtrl)
+	deploymentID := "deployment-id"
+	podID := "pod-id"
+	cases := map[string]struct {
+		containerPair               *containerPair
+		enrichedConnections         map[processListeningIndicator]timestamp.MicroTS
+		expectEntityLookupContainer expectFn
+		expectedStatus              *connStatus
+		expectedListeningIndicator  *processListeningIndicator
```
Comment (fredrb) [inline sensor/common/networkflow/manager/manager_impl_test.go:458]: Same idea about `assert` vs `expect` here.

**Review by fredrb — COMMENTED** [review]
Partial review, I didn't have time to look through all the tests yet. First of all, thanks a lot for putting the effort to cover this manager with unit tests!

I think the production code looks good, and the logic changed makes sense. 

With that said, I see some improvement to be made in the test structure, they are somewhat hard to follow, and there are some quick wins in terms of renaming some variables and extracting some functions that I suggest you to consider. 

As for testing internal functions vs public interfaces, give it a thought. I'm not so familiar with the code, so you can be the judge. Do you think that these function signatures and internal structures would benefit from refactors? If yes, adding these many tests that are calling them directly might make us shy away from making changes on them.

**Review by fredrb — COMMENTED** [review]
Another round. Did a pass on `sensor/common/networkflow/manager/manager_impl.go`

File: sensor/common/networkflow/manager/manager_impl.go [hunk 18]
```diff
@@ -273,11 +319,8 @@ func (m *networkFlowManager) enrichConnections() {
 func (m *networkFlowManager) enrichAndSend() {
 	currentConns, currentEndpoints := m.currentEnrichedConnsAndEndpoints()
 
-	updatedConns := computeUpdatedConns(currentConns, m.enrichedConnsLastSentState)
-	updatedEndpoints := computeUpdatedEndpoints(currentEndpoints, m.enrichedEndpointsLastSentState)
-
-	m.enrichedConnsLastSentState = currentConns
-	m.enrichedEndpointsLastSentState = currentEndpoints
+	updatedConns := computeUpdatedConns(currentConns, m.enrichedConnsLastSentState, &m.lastSentStateMutex)
+	updatedEndpoints := computeUpdatedEndpoints(currentEndpoints, m.enrichedEndpointsLastSentState, &m.lastSentStateMutex)
```
[the remark this answers is not in the document]
Comment (vikin91) [reply sensor/common/networkflow/manager/manager_impl.go:323]: The only argument against that proposal is when one of the functions in the critical section panics, then we would never unlock the mutex. Not sure if that is a true risk...

[the remark this answers is not in the document]
Comment (fredrb) [reply sensor/common/networkflow/manager/manager_impl.go:323]: You can extract those for lines to another function and `defer`. It's a nit though.

File: sensor/common/networkflow/manager/manager_impl_test.go (hunk 1 above)
Comment (fredrb) [reply sensor/common/networkflow/manager/manager_impl_test.go]: I prefer the code with the helpers. When looking at the test case, I feel like the code with the helpers shows just the important part, and it makes it easier to locate what changes from one test to the other. 

This whole thing may come down to personal preference, but since there are no strong signals, I'll leave mine here:

When it's a single test, I don't mind being more explicit, but if we have dozens of cases with the same boilerplate `.EXPECT()` parameters, then I'd prefer to move them to a helper function.

Take this one for example:

```go
"Incoming connection with successful lookup should not return a networkConnIndicator": {
	connPair:            createConnectionPair().incoming(),
	enrichedConnections: make(map[networkConnIndicator]timestamp.MicroTS),
	expectEntityLookupContainer: func() {
		mockEntityStore.EXPECT().LookupByContainerID(gomock.Any()).Times(1).DoAndReturn(func(_ any) (clusterentities.ContainerMetadata, bool) {
			return clusterentities.ContainerMetadata{
				DeploymentID: srcID,
			}, true
		})
	},
	expectEntityLookupEndpoint: func() {
		mockEntityStore.EXPECT().LookupByEndpoint(gomock.Any()).Times(1).DoAndReturn(func(_ any) []clusterentities.LookupResult {
			return []clusterentities.LookupResult{
				{
					Entity: networkgraph.Entity{
						ID: dstID,
					},
				},
			}
		})
	},
	expectedStatus: &connStatus{
		used: true,
	},
},
```


The things I'm interested in are:
- `createConnectionPair().incoming()`
- `clusterentities.ContainerMetadata` with `srcID`
- One `clusterentities.LookupResult` with `dstID` 
- And the assertion `connStatus.used == true`. 

But I have to squint my eyes and mentally remove all the boilerplate to see what's being changed from one test case to another. The alternative is much nicer, IMO:

```go
"Incoming connection with successful lookup should not return a networkConnIndicator": {
	connPair:            createConnectionPair().incoming(),
	enrichedConnections: make(map[networkConnIndicator]timestamp.MicroTS),
	expectEntityLookupContainer: expectEntityLookupContainerHelper(mockEntityStore, 1, clusterentities.ContainerMetadata{
		DeploymentID: srcID,
	}, true),
	expectEntityLookupEndpoint: expectEntityLookupEndpointHelper(mockEntityStore, 1, []clusterentities.LookupResult{
		{
			Entity: networkgraph.Entity{
				ID: dstID,
			},
		},
	}),
	expectedStatus: &connStatus{
		used: true,
	},
}
```

File: sensor/common/networkflow/manager/manager_impl_test.go [hunk 19]
```diff
@@ -179,7 +189,472 @@ func (suite *NetworkflowManagerTestSuite) TestAddNoOriginator() {
...
 	err := h.Process(networkInfoOpen, nowTimestamp, sequenceID)
-	suite.NoError(err)
+	s.NoError(err)
+
+	s.Len(h.endpoints, 1)
+}
+
+// endregion
+
+// region networkFlowManager tests
```
[the remark this answers is not in the document]
Comment (vikin91) [reply sensor/common/networkflow/manager/manager_impl_test.go:199]: They allow you to show/hide blocks of code in the IDE

Screenshot 2023-07-26 at 11 59 34
Screenshot 2023-07-26 at 11 59 41

File: sensor/common/networkflow/manager/manager_impl_test.go [hunk 20]
```diff
@@ -179,7 +189,472 @@ func (suite *NetworkflowManagerTestSuite) TestAddNoOriginator() {
...
+				srcEntity: networkgraph.EntityForDeployment(srcID),
+			},
+		},
+	}
+	for name, tCase := range cases {
+		s.Run(name, func() {
+			tCase.expectEntityLookupContainer.runIfSet()
+			tCase.expectEntityLookupEndpoint.runIfSet()
+			tCase.expectExternalLookup.runIfSet()
+			m.enrichConnection(tCase.connPair.conn, tCase.connPair.status, tCase.enrichedConnections)
```
[the remark this answers is not in the document]
Comment (fredrb) [reply sensor/common/networkflow/manager/manager_impl_test.go:388]: This can be postponed in favor of future refactors that would extract the `enrichConnection` function

File: sensor/common/networkflow/manager/manager_impl.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (fredrb) [reply sensor/common/networkflow/manager/manager_impl.go:323]: There is no problem in having these structs carry inconsistent data in case the connection breaks in-between each call. Also, the contention is not high enough to justify the performance improvement. Resolving.

File: sensor/common/networkflow/manager/manager_impl.go [hunk 21]
```diff
@@ -247,20 +253,60 @@ func (m *networkFlowManager) Capabilities() []centralsensor.SensorCapability {
...
 
+func (m *networkFlowManager) enrichConnections(tickerC <-chan time.Time) {
 	for {
 		select {
 		case <-m.done.WaitC():
 			return
-		case <-ticker.C:
+		case <-tickerC:
+			if !m.centralReady.IsDone() {
+				log.Debugf("sensor is in offline mode. Waiting %s to retry enriching and sending connections", tickerTime)
```
[the remark this answers is not in the document]
Comment (fredrb) [reply sensor/common/networkflow/manager/manager_impl.go]: Done

File: sensor/common/networkflow/manager/manager_impl_test.go (hunk 14 above)
Comment (fredrb) [reply sensor/common/networkflow/manager/manager_impl_test.go:214]: Skipping for now

File: sensor/common/networkflow/manager/manager_impl.go [hunk 22]
```diff
@@ -247,20 +253,60 @@ func (m *networkFlowManager) Capabilities() []centralsensor.SensorCapability {
...
 }
 
-func (m *networkFlowManager) Notify(common.SensorComponentEvent) {}
+func (m *networkFlowManager) Notify(e common.SensorComponentEvent) {
+	switch e {
+	case common.SensorComponentEventCentralReachable:
+		m.resetLastSentState()
+		m.centralReady.Signal()
+	case common.SensorComponentEventOfflineMode:
+		m.centralReady.Reset()
```
[the remark this answers is not in the document]
Comment (fredrb) [reply sensor/common/networkflow/manager/manager_impl.go:266]: Done
